File: android/app/src/main/java/ru/neuroimagen/arviewer/util/UniqueIdParser.kt
package ru.neuroimagen.arviewer.util

import android.net.Uri
import java.net.URI

/**
 * Single source of truth for parsing AR content unique_id from various inputs.
 *
 * Supported formats:
 * - Raw UUID: `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx`
 * - HTTPS link: `https://ar.neuroimagen.ru/view/{unique_id}`
 * - Custom scheme: `arv://view/{unique_id}`
 */
object UniqueIdParser {

    private const val EXPECTED_HOST = "ar.neuroimagen.ru"

    val UUID_REGEX = Regex(
        "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
    )

    /**
     * Extract unique_id from free-form user input (raw UUID or URL string).
     *
     * @return unique_id or null if format is not recognised
     */
    fun extractFromInput(input: String): String? {
        val trimmed = input.trim()
        if (UUID_REGEX.matches(trimmed)) return trimmed
        return try {
            parseFromUriString(trimmed)
        } catch (_: Exception) {
            null
        }
    }

    /**
     * Extract unique_id from a parsed [Uri] (deep link, QR code URL, etc.).
     *
     * @return unique_id or null if the URI does not match any known pattern
     */
    fun parseFromUri(uri: Uri): String? {
        return parseUriComponents(
            scheme = uri.scheme,
            host = uri.host,
            pathSegments = uri.pathSegments,
            path = uri.path,
        )
    }

    private fun parseFromUriString(value: String): String? {
        val uri = URI(value)
        return parseUriComponents(
            scheme = uri.scheme,
            host = uri.host,
            pathSegments = uri.path
                ?.trim('/')
                ?.split('/')
                ?.filter { it.isNotBlank() }
                .orEmpty(),
            path = uri.path,
        )
    }

    private fun parseUriComponents(
        scheme: String?,
        host: String?,
        pathSegments: List<String>,
        path: String?,
    ): String? {
        return when (scheme) {
            "arv" -> parseArvScheme(host, pathSegments, path)
            "https", "http" -> parseHttpScheme(host, pathSegments)
            else -> null
        }
    }

    /**
     * Returns true if [input] looks like a URL (starts with http/https/arv scheme).
     */
    fun looksLikeUrl(input: String): Boolean {
        return input.startsWith("http://") ||
                input.startsWith("https://") ||
                input.startsWith("arv://")
    }

    // ── Private helpers ──────────────────────────────────────────────

    private fun parseArvScheme(host: String?, pathSegments: List<String>, path: String?): String? {
        return if (host == "view") {
            pathSegments.firstOrNull()?.takeIf { UUID_REGEX.matches(it) }
                ?: path?.trimStart('/')?.takeIf { UUID_REGEX.matches(it) }
        } else {
            host?.takeIf { UUID_REGEX.matches(it) }
                ?: pathSegments.firstOrNull()?.takeIf { UUID_REGEX.matches(it) }
        }
    }

    private fun parseHttpScheme(host: String?, pathSegments: List<String>): String? {
        if (host != EXPECTED_HOST) return null
        val segments = pathSegments
        return if (segments.size >= 2 && segments[0] == "view") {
            segments[1].takeIf { UUID_REGEX.matches(it) }
        } else {
            null
        }
    }
}

File: ios/ARViewer/Utils/UniqueIdParser.swift
//
//  UniqueIdParser.swift
//  ARViewer
//
//  Извлечение unique_id из UUID, URL (https://ar.neuroimagen.ru/view/{id}), arv://view/{id}.
//

import Foundation

enum UniqueIdParser {
    private static let expectedHost = "ar.neuroimagen.ru"
    private static let uuidPattern = try! NSRegularExpression(
        pattern: "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
    )
    
    /// Проверяет, что строка — валидный UUID (формат).
    static func isValidUUID(_ s: String) -> Bool {
        let range = NSRange(s.trimmingCharacters(in: .whitespaces).startIndex..., in: s)
        return uuidPattern.firstMatch(in: s, range: range) != nil
    }
    
    /// Извлекает unique_id из ввода пользователя: сырой UUID или URL.
    static func extractFromInput(_ input: String) -> String? {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        if isValidUUID(trimmed) { return trimmed }
        guard let url = URL(string: trimmed) else { return nil }
        return parseFromURL(url)
    }
    
    /// Извлекает unique_id из URL (deep link или https).
    static func parseFromURL(_ url: URL) -> String? {
        if url.scheme?.lowercased() == "arv" {
            return parseArvScheme(url)
        }
        if url.scheme?.lowercased() == "https" || url.scheme?.lowercased() == "http" {
            return parseHTTPScheme(url)
        }
        return nil
    }
    
    static func looksLikeURL(_ input: String) -> Bool {
        let t = input.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.hasPrefix("http://") || t.hasPrefix("https://") || t.hasPrefix("arv://")
    }
    
    private static func parseArvScheme(_ url: URL) -> String? {
        guard url.host?.lowercased() == "view" else { return nil }
        let path = url.path.trimmingCharacters(in: "/")
        let first = path.split(separator: "/").first.flatMap(String.init) ?? path
        return first.isEmpty ? nil : (isValidUUID(first) ? first : nil)
    }
    
    private static func parseHTTPScheme(_ url: URL) -> String? {
        guard url.host?.lowercased() == expectedHost else { return nil }
        let comps = url.pathComponents.filter { $0 != "/" }
        if comps.count >= 2 && comps[0].lowercased() == "view" {
            let id = comps[1]
            return isValidUUID(id) ? id : nil
        }
        return nil
    }
}
